File: src/pyide/filelisting.py
import json
import os
import tornado.web
from pyide.configuration import SYS_PATH_PREPEND


class FileListing(tornado.web.RequestHandler):
    def get(self, path):
        result = []
        for root, dirs, files in os.walk((SYS_PATH_PREPEND + path)):
            padding_size = path.count('/')
            print('padding', padding_size, path, root, dirs, files)
            for d_dir in sorted(dirs):
                d_dir_path = os.path.join(root, d_dir).replace(SYS_PATH_PREPEND + '/', '')
                result.append(
                    '''<div class=folderlink ><img class="triange-img" style="transform: rotate(90deg);" src="/client/triangle.png"><span class="padding_{}">{}</span>\
<a href="/server/filelisting/{}">{}</a></div>'''.format(padding_size, '  ' * padding_size, d_dir_path, d_dir)
                )
            for f_file in sorted(f for f in files if not f.endswith('pyc')):
                f_file_path = os.path.join(root, f_file).replace(SYS_PATH_PREPEND + '/', '')
                result.append(
                    '''<div class=filelink   ><img class="triange-img" style="transform: rotate(90deg);"src="/client/blank.png"><span class="padding_{}">{}</span>\
<a href="/server/code/{}">{}</a></div>'''.format(padding_size, '  ' * padding_size, f_file_path, f_file))
            break
        self.write(json.dumps(result))

File: client/src/file-listing.ts
interface FListing {
    files: Array<string>,
    folders: Array<string>
}
export class FileListing {
    private fileListingElement: HTMLElement;
    public currentFileName: string = '';

    constructor() {
        this.fileListingElement = document.getElementById('filelisting');
        this.initListing().then();
    }

    createFolderElement(folderName: string){
        //TODO use template
        let divElement = document.createElement('div');
        divElement.className = 'folderlink';

        let img = document.createElement('img');
        img.className = 'triangle-img';
        img.style.transform = 'rotate(90deg)';
        img.src = '/client/resources/triangle.png';
        divElement.appendChild(img);

        let padding = Array.from(folderName.split('/')).length - 4;
        let span = document.createElement('span');
        span.textContent = '  '.repeat(padding);
        span.className = `padding_${padding}`;
        divElement.appendChild(span);


        let link = document.createElement('a');
        link.text = folderName.split('/').slice(-1)[0];
        link.href = folderName;
        divElement.appendChild(link);
        return divElement;
    }

    createFileElement(fileName: string){
        //TODO use template
        let divElement = document.createElement('div');
        divElement.className = 'filelink';

        let padding = Array.from(fileName.split('/')).length - 3;
        let span = document.createElement('span');
        span.textContent = '  '.repeat(padding);
        span.className = `padding_${padding}`;
        divElement.appendChild(span);

        let link = document.createElement('a');
        link.text = fileName.split('/').slice(-1)[0];
        link.href = fileName;
        divElement.appendChild(link);
        return divElement
    }

    async initListing() {
        let self = this;
        let response: FListing = await (await fetch('/server/filelisting')).json();
        response.folders.forEach(
            function (folderName: string) {
                self.fileListingElement.appendChild(self.createFolderElement(folderName));
            }
        );
        response.files.forEach(
            function (fileName: string) {
                self.fileListingElement.appendChild(self.createFileElement(fileName));
            }
        );
    }

    async get(event: MouseEvent) {
        let self = this;
        const target = <HTMLElement>event.target;
        let parentDiv = target.parentElement;
        const childNode0 = <HTMLElement>parentDiv.childNodes[0];

        if (parentDiv && childNode0.style.transform === 'rotate(180deg)') {
            childNode0.style.transform = 'rotate(90deg)';
            let childElement = parentDiv.lastChild;
            while (childElement) {
                if (['folderlink', 'filelink'].includes((<HTMLElement>childElement).className)) {
                    parentDiv.removeChild(childElement);
                    childElement = parentDiv.lastChild;
                } else {
                    break;
                }
            }
        } else if (childNode0.style.transform === 'rotate(90deg)') {
            childNode0.style.transform = 'rotate(180deg)';
            let response: FListing = await (await fetch((<HTMLAnchorElement>event.target).href)).json();
            response.folders.forEach(function (folderName: string) {
                parentDiv.appendChild(self.createFolderElement(folderName));
            });
            response.files.forEach(function (folderName: string) {
                parentDiv.appendChild(self.createFileElement(folderName));
            });
        } else {
            childNode0.style.transform = 'rotate(90deg)';
        }
    }

    async showFile(url: string) {
        this.currentFileName = url.split('/').slice(3).join('/');
        let response = await fetch(url);
        if (response.ok){
            return await response.json();
        }
        else{
            alert(`Server say: ${response.status} ${response.statusText}. On file name: ${this.currentFileName}`);
        }

    }
}
